ibis/udf/vectorized.py

```diff
@@ -38,7 +42,50 @@ def __call__(self, *args, **kwargs):
         # kwargs.
         @functools.wraps(self.func)
         def func(*args):
-            return self.func(*args, **kwargs)
+            # If cols are pd.Series, then we save and restore the index.
+            # Otherwise, we will reset the index to the natural index.
+            if hasattr(args[0], 'index'):
+                saved_index = args[0].index
+            else:
+                saved_index = None
+
+            result = self.func(*args, **kwargs)
+
+            if isinstance(self.output_type, dt.Struct):
+                if isinstance(result, pd.DataFrame) or (
+                    isinstance(result[0], (pd.Series, list, np.ndarray))
+                    and not any(
+                        isinstance(t, dt.Array) for t in self.output_type.types
+                    )
+                ):
+                    # We need to coerce_to_dataframe here because Spark
+                    # only allows DataFrame for struct output.
+                    result = coerce_to_dataframe(
+                        result, self.output_type.names
+                    )
+                    # Restore original index so that the result can later be
+                    # assigned to the original table with rows aligned
+                    if saved_index is not None:
+                        result.index = saved_index
+                    else:
+                        result = result.reset_index(drop=True)
```

Comment:
Why resetting index here?

Reply:
Yeah this might not actually be necessary.